sql/schemachanger: Enable ALTER TYPE on columns with DEFAULT or ON UPDATE expressions

spilchen opened the pull request:
Previously, complex column type alterations were blocked for columns with DEFAULT or ON UPDATE expressions. This change lifts that restriction. The primary challenge was managing computed columns, which cannot have DEFAULT or ON UPDATE expressions. Since a computed column is used to populate the new column during backfill, the DEFAULT or ON UPDATE expressions must only be added after the computed column is removed. New dependency rules now ensure that DEFAULT or ON UPDATE expressions are added only once the temporary computed expression is dropped.

To implement these rules correctly, I needed to determine whether the computed expression is part of an ALTER COLUMN operation. A new field was added to the `ColumnComputeExpression` element for tracking this usage.

Additionally, I've added checks to ensure that no column-dependent elements are left behind when dropping the old column. I reused debug code originally written for column drops, addressing some previously missed cases to clean up elements like `ColumnNotNull` and `ColumnComment`.

Epic: CRDB-25314
Closes #132909
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/134103)
<!-- Reviewable:end -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> Great work Matt! 

> Reviewed 107 of 107 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134103#-:-OAwwHnk-CNOsyBTbTTY:b-dyu8as)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134103#-OAwvMnQ69U82Uzm4JnD:-OAwvMnQ69U82Uzm4JnE:b-oiulhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3eded7e00f3819e2a1ef24b89b9c6344496a74/pkg/sql/schemachanger/scpb/elements.proto#L258)):*
> ```proto
>     // USING_EXPRESSION means it is part of the USING clause in an
>     // ALTER COLUMN .. TYPE operation.
>     USING_EXPRESSION = 1;
> ```

Nit: Might want to adjust the naming to be more direct, since I think this is only limited to ALTER TYPE.

In general this approach is okay, since its only needed while the schema change is active and the state is saved in the descriptor.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: Enable ALTER TYPE on columns with DEFAULT or ON UPDATE expressions

Previously, complex column type alterations were blocked on columns with
DEFAULT or ON UPDATE expressions. This change lifts that restriction.
The main challenge was handling computed columns, which cannot have
DEFAULT or ON UPDATE expressions. To populate the new column during
backfill, a temporary computed expression is used. New dependency rules
ensure that DEFAULT or ON UPDATE expressions are only added once the
transient computed expression is dropped.

To implement these rules correctly, I needed to determine whether the
computed expression is part of an ALTER COLUMN operation. A new field
was added to the `ColumnComputeExpression` element for tracking this
usage.

Additionally, I've added checks to ensure that no column-dependent
elements are left behind when dropping the old column. I reused debug
code originally written for column drops, addressing some previously
missed cases to clean up elements like `ColumnNotNull` and `ColumnComment`.

Epic: CRDB-25314
Closes #132909
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134103)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134103#-OAwvMnQ69U82Uzm4JnD:-OAxMHfm2WS0P_vYyw6V:b-g0fh2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3eded7e00f3819e2a1ef24b89b9c6344496a74/pkg/sql/schemachanger/scpb/elements.proto#L258)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: Might want to adjust the naming to be more direct, since I think this is only limited to ALTER TYPE.

In general this approach is okay, since its only needed while the schema change is active and the state is saved in the descriptor.
</blockquote></details>

Good suggestion! I’ve renamed it to `ALTER_TYPE_USING`. I hope that makes it clear and isn't too wordy.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR @fqazi 

bors r+

spilchen commented:
bors ping

spilchen commented:
bors r+
